pkg/apiserver/webhook/secret_webhook.go

```diff
@@ -0,0 +1,167 @@
+	Log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// Handle handles admission requests for Secret.
+func (v *SecretValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	v.Log.V(1).Info("Received admission webhook", "req", req, "operation", req.Operation)
+	switch req.Operation {
+	case admissionv1.Create:
+		return v.validateCreate(req)
+	case admissionv1.Update:
+		return v.validateUpdate(req)
+	case admissionv1.Delete:
+		return v.validateDelete(req)
+	default:
+		v.Log.Info("Invalid admission webhook request")
+	}
+	return admission.Allowed("")
+}
+
+// InjectDecoder injects the decoder.
+func (v *SecretValidator) InjectDecoder(d *admission.Decoder) error { //nolint:unparam
+	v.decoder = d
+	return nil
+}
+
+// allowSecretUpdate returns true only when Secret data key is unchanged.
+func (v *SecretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) bool {
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
+		return false
+	}
+	return true
+}
+
+// getCPABySecret returns nil only when the Secret is not used by any CloudProvideAccount CR,
+// otherwise the dependent CloudProvideAccount CR will be returned.
+func (v *SecretValidator) getCPABySecret(s *corev1.Secret) (error, *crdv1alpha1.CloudProviderAccount) {
+	// TODO: Figure out how to get all CRs.
+	cpaList := &crdv1alpha1.CloudProviderAccountList{}
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.Client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("%s, err:%v", errorMsg00001, err), nil
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.V(1).Info("Checking CloudProvideAccount", "Name", cpa.Name,
+			"Namespace", cpa.Namespace)
+		if cpa.Spec.AWSConfig != nil {
+			if cpa.Spec.AWSConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AWSConfig.SecretRef.Namespace == s.Namespace {
+				return nil, &cpa
+			}
+		}
+
+		if cpa.Spec.AzureConfig != nil {
+			if cpa.Spec.AzureConfig.SecretRef.Name == s.Name &&
+				cpa.Spec.AzureConfig.SecretRef.Namespace == s.Namespace {
+				return nil, &cpa
+			}
+		}
+	}
+	return nil, nil
+}
+
+func (v *SecretValidator) validateCreate(req admission.Request) admission.Response {
+	return admission.Allowed("")
+}
+
+func (v *SecretValidator) validateUpdate(req admission.Request) admission.Response {
+	newSecret := &corev1.Secret{}
+	err := v.decoder.Decode(req, newSecret)
+	if err != nil {
+		v.Log.Error(err, "Failed to decode Secret", "SecretValidator", req.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	oldSecret := &corev1.Secret{}
+	if req.OldObject.Raw != nil {
+		if err := json.Unmarshal(req.OldObject.Raw, &oldSecret); err != nil {
+			v.Log.Error(err, "Failed to decode old Secret", "SecretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	err, cpa := v.getCPABySecret(oldSecret)
+	if err != nil && strings.Contains(err.Error(), errorMsg00001) {
```

Comment:
you will return success case, if its an error and doesnt match a pattern. Please handle it too. I know now your code returns on that error, very easy to miss it in future

Reply:
sure. I get that.